tests/test_batch_norm.py

```diff
@@ -0,0 +1,287 @@
+        (7, 13),  # input_shape
+        13, 1e-05, 0.1, True,  # num_features, eps, momentum, affine,
+        False,  # track_running_statistics
+        4,  # passed to comm_split_fixture, required MPI ranks
+        id="distributed-batch-norm-affine-2d",
+        marks=[pytest.mark.mpi(min_size=4)]
+        )
+    )
+
+parametrizations_affine.append(
+    pytest.param(
+        np.arange(0, 8), [1, 2, 2, 2],  # P_x_ranks, P_x_shape,
+        (7, 13, 11, 3),  # input_shape
+        13, 1e-05, 0.1, True,  # num_features, eps, momentum, affine,
+        False,  # track_running_statistics
+        8,  # passed to comm_split_fixture, required MPI ranks
+        id="distributed-batch-norm-affine-4d",
+        marks=[pytest.mark.mpi(min_size=8)]
+        )
+    )
+
+parametrizations_affine.append(
+    pytest.param(
+        np.arange(0, 12), [1, 3, 2, 2],  # P_x_ranks, P_x_shape,
+        (7, 13, 11, 3),  # input_shape
+        13, 1e-05, 0.1, True,  # num_features, eps, momentum, affine,
+        False,  # track_running_statistics
+        12,  # passed to comm_split_fixture, required MPI ranks
+        id="distributed-batch-norm-affine-4d-many-ranks",
+        marks=[pytest.mark.mpi(min_size=12)]
+        )
+    )
+
+parametrizations_non_affine.append(
+    pytest.param(
+        np.arange(0, 4), [1, 2, 2],  # P_x_ranks, P_x_shape,
+        (7, 13, 11),  # input_shape
+        13, 1e-05, 0.1, False,  # num_features, eps, momentum, affine,
+        False,  # track_running_statistics
+        4,  # passed to comm_split_fixture, required MPI ranks
+        id="distributed-batch-norm-no-affine-feature",
+        marks=[pytest.mark.mpi(min_size=4)]
+        )
+    )
+
+parametrizations_non_affine.append(
+    pytest.param(
+        np.arange(0, 4), [1, 2, 2],  # P_x_ranks, P_x_shape,
+        (7, 13, 11),  # input_shape
+        13, 1e-05, 0.1, False,  # num_features, eps, momentum, affine,
+        True,  # track_running_statistics
+        4,  # passed to comm_split_fixture, required MPI ranks
+        id="distributed-batch-norm-no-affine-feature-track-running",
+        marks=[pytest.mark.mpi(min_size=4)]
+        )
+    )
+
+
+@pytest.mark.parametrize("P_x_ranks, P_x_shape,"
+                         "input_shape,"
+                         "num_features, eps, momentum, affine,"
+                         "track_running_stats,"
+                         "comm_split_fixture",
+                         parametrizations_affine,
+                         indirect=["comm_split_fixture"])
+def test_batch_norm_with_training(barrier_fence_fixture,
+                                  P_x_ranks, P_x_shape,
+                                  input_shape,
+                                  num_features, eps, momentum, affine,
+                                  track_running_stats,
+                                  comm_split_fixture):
+
+    from distdl.backends.mpi.partition import MPIPartition
+
+    torch.manual_seed(0)
```

Comment:
Is forcing the seed necessary here?  Your distributed network shares out the copy of the data so the seed should be allowed to vary?

Reply:
Because I initialize the inputs randomly, and there is a margin of error that is difficult to quantify (especially due to the squaring), _sometimes_ tests will fail at this error threshold, but it is a rare event. Maybe we can randomize with several different seeds? Or perhaps is there a way to be more fine-grained with the error threshold?